teps/0076-array-result-types.md

```diff
@@ -0,0 +1,321 @@
+  results. The proposal to add dictionary is mostly motivated by use cases for dictionary
+  results (vs params) - but today we only allow string results. It seems confusingly
+  inconsistent to imagine adding dictionary results but not array results (even though
+  we support array params)
+* Without indexing, arrays will only work with arrays. This means you cannot combine arrays with Tasks that
+  only take string parameters and this limits the interoperabiltiy between Tasks.
+
+### Goals
+
+* Add two missing pieces of array support
+  * As results
+  * Indexing into arrays
+* Take a step in the direction of allowing Tasks to have even more control
+  over their params and results ([see pipelines#1393](https://github.com/tektoncd/pipeline/issues/1393))
+* Be consistent in our treatment of arrays in Pipelines and Tasks and other types, e.g.
+  dictionaries (TEP-0075)
+
+### Non-Goals
+
+* Adding complete [JSONPath](https://tools.ietf.org/id/draft-goessner-dispatch-jsonpath-00.html)
+  syntax support
+* Adding support for nesting, e.g. array params where the items in the array are themselves arrays
+  (no reason we can't add that later but trying to keep it simple for now)
+
+### Use Cases
+
+1. Looping (i.e. providing values for an interface such as the task-loop custom task
+   [iterateParam](https://github.com/tektoncd/experimental/tree/main/task-loops#specifying-the-iteration-parameter)
+   (see the workaround added in [experimental#713](https://github.com/tektoncd/experimental/pull/713) which
+   splits a string). For example, in tektoncd/plumbing we have a directory of
+   [images](https://github.com/tektoncd/plumbing/tree/main/tekton/images). Each directory contains a Dockerfile -
+   in a Pipeline that builds all images in this repo, we would want to: 1) determine which Dockerfiles need to be built
+   and 2) for each of those Dockerfiles, build the image. Looping support would allow us to use a Task which builds
+   just one image (e.g. [the kaniko task](https://github.com/tektoncd/catalog/tree/main/task/kaniko/0.4)), and run it
+   for each Dockerfile, instead of having to the task to take multiple Dockerfiles as input. This missing piece of this
+   is supporting the Task which determines with Dockerfiles to build and provides them as an array (instead of
+   encoded in a string).
+
+   ```
+   filter-dockerfiles
+        |
+        | result: list of dockerfiles
+        v
+   (for each: filter-dockerfiles result)
+      kaniko build
+   ```
+
+## Requirements
+
+* Must be possible for a result array to be empty
+* Must be possible to use the array results of one Task in a Pipeline as the
+  param of another Task in the pipeline which has an array param
+* Must be possible to use one specific value in an array result of one Task
+  in a Pipeline as the param of another Task in the pipeline which has a string param
+  * If there is no value at the specified index, the Pipeline execution should fail
```

Comment:
I don't think the benefit of the use case provided is addressed by these requirements. The key missing piece for me is dynamism. Already knowing the size of your array upfront is great and all but I think it is worth exploring with this effort how we can support variable sized arrays and perhaps even task generation based on some sort of template.

Reply:
Dynamism here is supported using task loops, since you don't need to know the size of the array there, you just pass the whole array and the custom controller does the rest.
Eventually we should get loops into the main API to support what you're asking for without the need for custom tasks.

I'm not sure tbh what the use case for indexing is?

Reply:
After the discussion on https://github.com/tektoncd/pipeline/issues/4097, I think I see a good use case for indexing - i.e. especially for results, there are cases where it makes sense of processing an array of results instead of feeding it to a loop of tasks directly.

This will require being able to know the size of the array though, so it feels like indexing and length should go together?

Reply:
I'm assuming this is addressed via the conversation at https://github.com/tektoncd/community/pull/477#discussion_r673508137 but plz re-open if not!